src/pubd/repository.rs

```diff
@@ -764,20 +767,172 @@ pub struct RrdpServer {
+    /// existing, but unpublished, updates applied to it. So, they may wish
+    /// to make further updates before the staged updates are actually published.
+    ///
+    /// But if they do, then the new elements should be merged into this existing
+    /// StagedElements so that it can be applied to the current snapshot. This
+    /// way we can have a single RRDP delta that combines all changes.
+    ///
+    /// This operation is not entirely trivial as we need to make sure that
+    /// any hashes in updates and withdraws after merge match the current
+    /// snapshot, i.e. the new snapshot could refer to hashes of not yet
+    /// published objects. We may also simply "forget" objects that were
+    /// staged for publication, and then withdrawn, without ever have been
+    /// published.
+    ///
+    /// Also note (again) that all changes have been verified before when
+    /// this is called. That means that while certain corner cases could be
+    /// problematic (e.g. double withdraw of an object), we will largely
+    /// ignore such issues here. We need to do this, because we get these
+    /// changes as write-ahead-log (WAL) changes and therefore applying
+    /// them is not allowed to fail.
+    fn merge_new_elements(&mut self, elements: DeltaElements) {
+        let (publishes, updates, withdraws) = elements.unpack();
+
+        for pbl in publishes {
+            let uri = pbl.uri().clone();
+            match self.0.get_mut(&uri) {
+                Some(DeltaElement::Publish(_)) => {
+                    // This should never happen as the new publish would have
+                    // been an update instead. This will have been checked
+                    // at this point. Still, rather than failing which is
+                    // not allowed in this code path, we can use the publish
+                    // instead of the staged publish.
+                    self.0.insert(uri, DeltaElement::Publish(pbl));
+                }
+
+                Some(DeltaElement::Update(staged_update)) => {
+                    // This should never happen as the new publish would have
+                    // been an update instead. This will have been checked
+                    // at this point. Still, rather than failing which is
+                    // not allowed in this code path, we can update the
+                    // existing update with the content of the new publish
+                    // element.
+                    staged_update.with_updated_content(staged_update.base64().clone());
+                }
+                Some(DeltaElement::Withdraw(staged_withdraw)) => {
+                    // A new publish that follows a withdraw for the same URI should be
+                    // an Update of the original file.
+                    let hash = *staged_withdraw.hash();
+                    let update = UpdateElement::new(uri.clone(), hash, pbl.base64().clone());
```

Comment:
What is "hash" on UpdateElement meant to be the hash of? The thing being updated, in which case maybe this is correct? Or the update being supplied, in which case it looks a bit odd to take the hash of the previous withdraw and associate it with the new update... ?

Reply:
The hash matches the thing being updated - i.e. the state this StagedElements will be applied to, so we always need to preserve the hash when merging.